Refactor and Extend `system-info` Handler to Obtain Generic System Information

ejoerns opened the pull request:
This cleans up a bit of the handler code, fixes some issues and prepares for more.

The actual focus of this is to extend the already-existing `system-info` handler (that had no real purpose yet) with the capability to add and collect custom information. For now, these informations are made available in pre- and post-install handler env variables only, but there will be follow-up work to provide them via HTTP headers.

Preparation is related to #1114

jluebbe commented on src/context.c:
```diff
@@ -225,29 +225,28 @@ static gboolean get_system_info_from_handler(GError **error)
 	}
 
-	return TRUE;
+	return g_steal_pointer(&vars);
```
Can't we assign context->

ejoerns commented on src/install.c:
```diff
@@ -1138,7 +1137,7 @@ gboolean do_install_bundle(RaucInstallArgs *args, GError **error)
 	if (r_context()->config->preinstall_handler) {
 		g_message("Starting pre install handler: %s", r_context()->config->preinstall_handler);
-		res = launch_and_wait_handler(args, bundle->mount_point, r_context()->config->preinstall_handler, bundle->manifest, target_group, &ierror);
+		res = launch_and_wait_handler(args, bundle->mount_point, r_context()->config->preinstall_handler, bundle->manifest, NULL, target_group, &ierror);
```
These are *manifest* settings for the *full custom handler*. I don't really see how these should have any relation to the *system.conf*-defined *pre-/post-install handlers*. They do not have anything in common (except that they, unfortunately, share some code and naming). Thus I would argue extra args should not be passed to 'normal' handlers. If one actually wants to pass custom arguments from the manifest, I guess a more explicit handling would be better.

ejoerns commented on src/install.c:
```diff
@@ -1158,8 +1168,10 @@ gboolean do_install_bundle(RaucInstallArgs *args, GError **error)
+		g_auto(GStrv) env = NULL;
 		g_message("Starting post install handler: %s", r_context()->config->postinstall_handler);
-		res = launch_and_wait_handler(args, bundle->mount_point, r_context()->config->postinstall_handler, bundle->manifest, NULL, target_group, &ierror);
+		env = prepare_environment(bundle->mount_point, bundle->manifest, target_group);
```
Yes, better. Will fix.

ejoerns commented on src/install.c:
```diff
@@ -551,12 +551,24 @@ static gboolean verify_compatible(RaucInstallArgs *args, RaucManifest *manifest,
+
+		g_hash_table_iter_init(&iter, r_context()->system_info);
+		while (g_hash_table_iter_next(&iter, (gpointer*) &key, (gpointer*) &value)) {
+			envp = g_environ_setenv(envp, key, value, TRUE);
```
Ok

ejoerns commented on src/install.c:
```diff
@@ -770,6 +776,7 @@ static gboolean launch_and_wait_custom_handler(RaucInstallArgs *args, gchar* bun
 	GError *ierror = NULL;
 	g_autofree gchar* handler_name = NULL;
 	g_autoptr(GString) handler_args = NULL;
+	gchar **env = NULL;
```
Yes, will fix.

ejoerns commented on src/install.c:
```diff
@@ -645,23 +645,15 @@ static gboolean launch_and_wait_handler(RaucInstallArgs *args, gchar *update_sou
 			handlelaunch, &ierror,
 			handler_name,
-			handler_args->str,
+			handler_args,
```
Here we go..

ejoerns commented:
Changes:

* added helper function to add GStrv to GPtrArray
* fixed argv parsing and handling
* added and fixed documentation on full custom handler
* slightly extended install tests for handler args
* updated documentation for changes in system-info handler
* debug output for handler subprocess launch
* call `prepare_environment` only once in `do_install_bundle()`
* do not override env variables in `add_sytem_environment()`
* add error check for using `[handler]` `args` without `filename`

jluebbe commented on test/context.c:
```diff
@@ -67,6 +67,30 @@ static void test_bootslot_no_bootslot(void)
+
+	r_context_clean();
+}
+
+
```
```suggestion

```

jluebbe commented on test/install.c:
```diff
@@ -1496,7 +1497,7 @@ int main(int argc, char *argv[])
 			.message_needles = memdup((&(const gchar *[]) {
 				"Checking and mounting bundle...",
-				"Debug: --dummy1 --dummy2",
+				"Debug: arg1 arg2 --dummy1 --dummy2",
```
How do the multiple sources of arguments interact? That should be documented.

ejoerns replied:
I have not found a good way to provide more verbose argument descriptions. Maybe the man page or a separate reference chapter would be a proper place. For now I have added a short hint in the manifest reference about arguments.
I have now also clarified the purpose of ``--handler-args``  in the help output a bit.

jluebbe commented on src/install.c:
```diff
@@ -608,61 +634,65 @@ static void prepare_environment(GSubprocessLauncher *launcher, gchar *update_sou
 }
 
-static gboolean launch_and_wait_handler(RaucInstallArgs *args, gchar *update_source, gchar *handler_name, RaucManifest *manifest, GHashTable *target_group, GError **error)
+static gboolean launch_and_wait_handler(RaucInstallArgs *args, gchar *handler_name, gchar **handler_argv, gchar **setenv, GError **error)
```
Perhaps use `const` to indicate that these should not be changed?

ejoerns replied:
For the `handler_argv`? I guess I even had it first like this, let's see where I have to cast around.

jluebbe replied:
If it is too much hassle due to glib function prototypes, skip this.

jluebbe commented on include/utils.h:
```diff
@@ -68,6 +68,20 @@ G_GNUC_WARN_UNUSED_RESULT;
+ * Adds elements of a zero-terminated GStrv/gchar** to an existing GPtrArray
+ *
+ * @param ptrarray GPtrArray to add to
+ * @param argv arguments to add
```
```suggestion
 * @param argvp arguments to add
```

jluebbe commented on src/context.c:
```diff
@@ -293,7 +292,8 @@ static gboolean r_context_configure_target(GError **error)
 	}
 
 	if (context->config->systeminfo_handler) {
```
Should we create an empty `GHashTable` if the system info handler is not set? That would avoid extra checks later.

ejoerns replied:
Could help. I'll give it a try.

jluebbe commented on docs/reference.rst:
```diff
@@ -561,12 +561,40 @@ A valid RAUC manifest file must be named ``manifest.raucm``.
 **[handler] section**
 
+The ``handler`` section refers to the *full custom handler* that allows
+to fully replace the default RAUC update process.
```
Add a reference to https://rauc.readthedocs.io/en/latest/using.html#full-custom-update.

ejoerns replied:
This hardly gave any additional info. I have added the link but also slightly extended the chapter that it refers to

jluebbe commented on src/install.c:
```diff
@@ -608,35 +611,37 @@ static void prepare_environment(GSubprocessLauncher *launcher, gchar *update_sou
 }
 
-static gboolean launch_and_wait_handler(RaucInstallArgs *args, gchar *update_source, gchar *handler_name, RaucManifest *manifest, gchar **handler_argv, GHashTable *target_group, GError **error)
+static gboolean launch_and_wait_handler(RaucInstallArgs *args, gchar *handler_name, gchar **handler_argv, gchar **setenv, GError **error)
```
`setenv` is a function from `stdlib.h`, so we should not use it as a parameter name.

ejoerns commented on include/utils.h:
```diff
@@ -68,6 +68,20 @@ G_GNUC_WARN_UNUSED_RESULT;
+ * @param argv arguments to add
+ * @param copy whether to just add the pointer (FALSE) or copy the underlying data (TRUE)
+ */
+static inline void r_g_ptr_array_addv(GPtrArray *ptrarray, gchar **argvp, gboolean copy)
```
to me too. I just left the `g`  to reflect that it operates on **G**PtrArray. But maybe this is clear from the context and `r_ptr_array_add` is sufficient.

ejoerns commented on docs/reference.rst:
```diff
@@ -559,12 +562,40 @@ A valid RAUC manifest file must be named ``manifest.raucm``.
+  Arguments to pass to the full custom handler, such as
+  ``args=--setup --verbose``
+
+  .. note:: Up to RAUC v1.9, these arguments were also implicitly passed
+     to handlers defined in the system.conf.
+     This behavior was fixed/removed now.
+     If someone made use of this undocumented behavior and requires this to
+     work, please file an `issue
+     <https://github.com/rauc/rauc/issues/new/choose>`_.
```
`fixed/removed in v1.10`, but otherwise fine

ejoerns commented on include/utils.h:
```diff
@@ -68,6 +68,20 @@ G_GNUC_WARN_UNUSED_RESULT;
+ */
+static inline void r_g_ptr_array_addv(GPtrArray *ptrarray, gchar **argvp, gboolean copy)
+{
+	for (gchar **addarg = argvp; *addarg != NULL; addarg++) { \
+		g_ptr_array_add(ptrarray, copy ? g_strdup(*addarg) : *addarg); \
```
Oh, I made it to obvious that I started with a macro I guess :smirk:

ejoerns commented:
Another round. Changes:

* added link to full custom update in using.rst and extended the description found there
* renamed `r_g_ptr_array_addv` to `r_ptr_array_addv`
* changed help text of ``--handler-args`` to contain the term "full custom handler"
* always initialize `context->system_info` (and remove now useless NULL check)
* added method documentation for `prepare_environment()` and `launch_and_wait_handler`
* renamed `setenv` argument to `override_env` to avoid naming collisions
* further minor fixes mentioned in review above
